dwh/query_lib.py

```diff
@@ -0,0 +1,278 @@
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""This is the main higher level library to query FHIR resources.
+
+The public interface of this library is intended to be independent of the actual
+query engine, e.g., Spark, SQL/BigQuery, etc. The only exception is a single
+function that defines the source of the data.
+"""
+
+# See https://stackoverflow.com/questions/33533148 why this is needed.
+from __future__ import annotations
+from enum import Enum
+from typing import List
+import pandas
+from pyspark import SparkConf
+from pyspark.sql import SparkSession, DataFrame
+import pyspark.sql.functions as F
+
+import common
+
+
+class Runner(Enum):
+  SPARK = 1
+  BIG_QUERY = 2
+  #FHIR_SERVER = 3
+
+
+def patient_query_factory(runner: Runner, data_source: str) -> PatientQuery:
+  """Returns the right instance of `PatientQuery` based on `data_source`.
+
+  Args:
+    runner: The runner to use for making data queries
+    data_source: The definition of the source, e.g., directory containing
+      Parquet files or a BigQuery dataset.
+
+  Returns:
+    The created instance.
+
+  Raises:
+    ValueError: When the input `data_source` is malformed or not implemented.
+  """
+  if runner == Runner.SPARK:
+    return _SparkPatientQuery(data_source)
+  if runner == Runner.BIG_QUERY:
+    return _BigQueryPatientQuery(data_source)
+  raise ValueError('Query engine {} is not supported yet.'.format(runner))
+
+
+class _ObsConstraints():
+
+  def __init__(self, code: str, values: List[str] = None,
+       min_value: float = None, max_value: float = None,
+       min_time: str = None, max_time: str = None) -> None:
+     self._code = code
+     self._values = values
+     self._min_time = min_time
+     self._max_time = max_time
+     self._min_value = min_value
+     self._max_value = max_value
+
+  @staticmethod
+  def time_constraint(min_time: str = None, max_time: str = None):
+    if not min_time and not max_time:
+      return 'TRUE'
+    cl = []
+    if min_time:
+      cl.append('dateTime >= "{}"'.format(min_time))
+    if max_time:
+      cl.append('dateTime <= "{}"'.format(max_time))
+    return ' AND '.join(cl)
+
+  def sql(self) -> str:
+    """This creates a constraint string with WHERE syntax in SQL.
+
+    All of the observation constraints specified by this instance are joined
+    together into an `AND` clause.
+    """
+    cl = [self.time_constraint(self._min_time, self._max_time)]
+    cl.append('coding.code="{}"'.format(self._code))
+    if self._values:
+      codes_str = ','.join(['"{}"'.format(v) for v in self._values])
+      cl.append('value.string IN ({})'.format(codes_str))
```

Comment:
Thanks for adding the ability to filter based on codes/strings. Take a look at this scenario

A patient comes to the clinic on 2021/04/03 and is marked as on ART i.e., `(ART_PLAN = CONTINUE_REGIMEN`  then same patient comes to the clinic on `2021/04/25` and the clinician stops ART i.e., `(ART_PLAN = STOP ART)` - will this patient be counted as on ART given the constraint is `coding.code="ART_PLAN_CODE" AND value.string IN ("CONTINUE_REGIMEN_CODE","START_REGIMEN_CODE"))`

Reply:
This constraint is for choosing obs in the Spark space, if `STOP_ART` is relevant to the indicator that is being calculated, then the set of values should be `("STOP_ART", "CONTINUE_REGIMEN_CODE","START_REGIMEN_CODE")`. With the new aggregation schema, the pivoting can be done in Panadas space on `valueCode` column after filtering `ART_PLAN` codes. So in this example, since `STOP_ART` `max_date` would be greater than that of `CONTINUE_REGIMEN_CODE` and `START_REGIMEN_CODE` I think this should not be counted (if this is what the indicator says).

Reply:
thanks for elaborating on this? This makes sense!